test/distributed/_composable/test_composability/test_pp_composability.py

```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
 
         torch.distributed.destroy_process_group()
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(4)
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "Test requires 4+ GPUs")
+    def test_pp_and_dcp(self):
+        """
+        Test that pipeline parallelism and distributed checkpointing can be used together and
+        with saved correct FQNs
+        """
+
+        class AppState(Stateful):
+            def __init__(self, model, optimizer):
+                self.model = model
+                self.optimizer = optimizer
+
+            def state_dict(self):
+                # this line automatically manages FSDP FQN's, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
+                model_state_dict, optimizer_state_dict = get_state_dict(
+                    self.model, self.optimizer
+                )
+                return {"model": model_state_dict, "optim": optimizer_state_dict}
+
+            def load_state_dict(self, state_dict):
+                # sets our state dicts on the model and optimizer, now that we've loaded
+                set_state_dict(
+                    self.model,
+                    self.optimizer,
+                    model_state_dict=state_dict["model"],
+                    optim_state_dict=state_dict["optim"],
+                )
+
+        class PPModelChunk(nn.Module):
+            def __init__(self, layers: nn.ModuleDict, start_index: int, end_index: int):
+                super().__init__()
+                # Filter layers based on start_index and end_index
+                self.layers = nn.ModuleDict(
+                    {str(i): layers[str(i)] for i in range(start_index, end_index)}
+                )
+
+            def forward(self, x):
+                for layer in self.layers.values():
+                    x = layer(x)
+                return x
+
+        device = torch.device("cuda", self.device)
+        torch.cuda.set_device(self.device)
+        store = torch.distributed.FileStore(self.file_name, self.world_size)
+        torch.distributed.init_process_group(
+            backend="nccl",
+            store=store,
+            rank=self.rank,
+            world_size=self.world_size,
+            device_id=device,
+        )
+        # create "entire model"
+        total_layers = 8
+        dim = 10
+        full_model = nn.ModuleDict(
+            {f"{i}": MLPModule(dim) for i in range(total_layers)}
+        )
+        # Calculate start and end indices based on rank
+        start_index = self.rank * 2
+        end_index = start_index + 2
+        pp_model = PPModelChunk(full_model, start_index, end_index)
+
+        pp_model.to(self.device)
+        opt = torch.optim.Adam(pp_model.parameters(), lr=0.1)
+
+        # Manually create a temporary directory
+        # Create a temporary directory on rank 0 and broadcast its path
```

Comment:
This can be tricky sometimes. Like rank0 creates the folder but is not seen by other ranks on the same machine. I would suggest to use `from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir`.

Reply:
You can directly use `self.temp_dir` when using @with_temp_dir. See example here: https://github.com/pytorch/pytorch/blob/fdfd4c50bad8b7d12416d45233ad990c45cf7ef9/test/distributed/checkpoint/test_format_utils.py#L50

Reply:
thanks, updated to use `@with_temp_dir`